File: 1.two-sum.java
/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // int[] result = new int[2];
        HashMap<Integer, Integer> history = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            if (history.containsKey(target - nums[i])) {
                // result[0] = i;
                // result[1] = history.get(target-nums[i]);
                // return result;
                return new int[] { i, history.get(target - nums[i]) };
            }
            history.put(nums[i], i);
        }
        return new int[] { 0, 0 };
    }
}
// @lc code=end

File: 1.two-sum.rs
/*
 * @lc app=leetcode id=1 lang=rust
 *
 * [1] Two Sum
 *
 * https://leetcode.com/problems/two-sum/description/
 *
 * algorithms
 * Easy (49.23%)
 * Likes:    45690
 * Dislikes: 1487
 * Total Accepted:    9.5M
 * Total Submissions: 19.1M
 * Testcase Example:  '[2,7,11,15]\n9'
 *
 * Given an array of integers nums and an integer target, return indices of the
 * two numbers such that they add up to target.
 *
 * You may assume that each input would have exactly one solution, and you may
 * not use the same element twice.
 *
 * You can return the answer in any order.
 *
 *
 * Example 1:
 *
 *
 * Input: nums = [2,7,11,15], target = 9
 * Output: [0,1]
 * Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 *
 *
 * Example 2:
 *
 *
 * Input: nums = [3,2,4], target = 6
 * Output: [1,2]
 *
 *
 * Example 3:
 *
 *
 * Input: nums = [3,3], target = 6
 * Output: [0,1]
 *
 *
 *
 * Constraints:
 *
 *
 * 2 <= nums.length <= 10^4
 * -10^9 <= nums[i] <= 10^9
 * -10^9 <= target <= 10^9
 * Only one valid answer exists.
 *
 *
 *
 * Follow-up: Can you come up with an algorithm that is less than O(n^2) time
 * complexity?
 */

// @lc code=start
use std::{collections::HashMap, iter};
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut memo = HashMap::new();

        for (idx, num) in nums.iter().enumerate() {
            let target = target - num;
            if (memo.contains_key(&target)) {
                return vec![*memo.get(&target).unwrap() as i32, idx as i32];
            }
            memo.insert(*num, idx as i32);
        }
        return vec![];
    }
}
// @lc code=end
